packages/doenetml-worker-rust/lib-doenetml-core/src/core/graph_node_lookup.rs

```diff
@@ -0,0 +1,167 @@
+    strings: Vec<Option<T>>,
+    props: Vec<Option<T>>,
+    states: Vec<Option<T>>,
+    queries: Vec<Option<T>>,
+    virtuals: Vec<Option<T>>,
+}
+
+impl<T> GraphNodeLookup<T> {
+    pub fn new() -> Self {
+        Self {
+            components: Vec::new(),
+            strings: Vec::new(),
+            props: Vec::new(),
+            states: Vec::new(),
+            queries: Vec::new(),
+            virtuals: Vec::new(),
+        }
+    }
+
+    pub fn keys(&self) -> impl Iterator<Item = GraphNode> {
+        let all_nodes = self
+            .components
+            .iter()
+            .enumerate()
+            .filter_map(|(i, item)| item.as_ref().map(|_| GraphNode::Component(i)))
+            .chain(
+                self.strings
+                    .iter()
+                    .enumerate()
+                    .filter_map(|(i, item)| item.as_ref().map(|_| GraphNode::String(i))),
+            )
+            .chain(
+                self.props
+                    .iter()
+                    .enumerate()
+                    .filter_map(|(i, item)| item.as_ref().map(|_| GraphNode::Prop(i))),
+            )
+            .chain(
+                self.states
+                    .iter()
+                    .enumerate()
+                    .filter_map(|(i, item)| item.as_ref().map(|_| GraphNode::State(i))),
+            )
+            .chain(
+                self.queries
+                    .iter()
+                    .enumerate()
+                    .filter_map(|(i, item)| item.as_ref().map(|_| GraphNode::Query(i))),
+            )
+            .chain(
+                self.virtuals
+                    .iter()
+                    .enumerate()
+                    .filter_map(|(i, item)| item.as_ref().map(|_| GraphNode::Virtual(i))),
+            )
+            .collect::<Vec<_>>();
+
+        all_nodes.into_iter()
+    }
+
+    pub fn iter(&self) -> impl Iterator<Item = (GraphNode, &T)> {
+        self.keys().map(|node| (node, self.get_tag(&node).unwrap()))
+    }
+}
+
+impl<T> Taggable<GraphNode, T> for GraphNodeLookup<T> {
+    fn get_tag(&self, node: &GraphNode) -> Option<&T> {
+        match node {
+            GraphNode::Component(idx) => self.components.get(*idx).and_then(|x| x.as_ref()),
+            GraphNode::String(idx) => self.strings.get(*idx).and_then(|x| x.as_ref()),
+            GraphNode::Prop(idx) => self.props.get(*idx).and_then(|x| x.as_ref()),
+            GraphNode::State(idx) => self.states.get(*idx).and_then(|x| x.as_ref()),
+            GraphNode::Query(idx) => self.queries.get(*idx).and_then(|x| x.as_ref()),
+            GraphNode::Virtual(idx) => self.virtuals.get(*idx).and_then(|x| x.as_ref()),
+        }
+    }
+    fn set_tag(&mut self, node: GraphNode, tag: T) {
+        let idx = node.idx();
+        match node {
+            GraphNode::Component(_) => {
+                if idx >= self.components.len() {
+                    self.components.resize_with(idx + 1, || None);
```

Comment:
Why `resize_with()` rather than `resize()`?

Reply:
`resize` requires `T` to be clone, which would result in a lot more trait bounds...